### ISSUE TITLE
How to parse the .txt data file?

### ISSUE THREAD

**Issue by SoonSYJ** [IRC-SPHERE/elsts2017-temperature-data#1]
Hi, 
I am confusing about how to get the node's "drift" and "temperature" data. As the "readme" file said that the temperature data are located in the .csv file and the drift are located in the .txt file. 
The problem is when I looking into the .txt file, I try to search the keyword "drift" and the number of the results (~100) are quite small than the data length (~8000) in the .csv file.
Could you give me some advice about parsing the .csv and .txt file. 
THanks a lot

**Comment by atiselsts** [comment]
Hi SoonSYJ ,

the log files (.txt) are just the raw logging output of nodes.  There are three types of lines you need to extract:

```
1494298887.2 > TSCH: {asn-0.575016 link-0-0-0-0 ch-26} drift -65
1494298887.2 > TSCH: {asn-0.575016 link-0-0-0-0 ch-26} bc-0-0 35 rx 131, dr -2 (-24562), edr 2 (24562)
1494298887.4 > TSCH: {asn-0.57502b link-0-0-0-0 ch-20} bc-0-0 35 rx 131, edr 1 (-299)
```

- The `1494298887.2` is server time, in seconds.
- `asn-0.575016` is the ASN slot number, in the 5 byte hex notation.
- `drift -65` is the drift estimate of the sensor node. It's printed in custom units (1/1024 of a microsecond). It is only printed after its updated, and it's only updated on synchronization events.
- `dr -2 (-24562), edr 2 (24562)` shows that EB packet was received and used for synchronization.  `24562` is the corrected synchronization error, in 1/1024 of a microsecond. Time correction also triggers update of the `drift` value (which was already printed, in the line before this one).
- `edr 1 (-299)` shows that EB packet was received, but not used for syncronization. `-299` is the estimated error.

In the experiment, beacons were sent around once per second, but sync only once per several minutes (at least after the initial learning period). So we have few synchronization events, definitely much less than temperature sensing events.

We use these regular expressions to parse the lines (the server time is ignored):
```
regex_drift_value = ".*TSCH: {asn-([0-9a-f])\.([0-9a-f]+) .*} drift (-?[0-9]+)$"
regex_edr_line = ".*TSCH: {asn-([0-9a-f])\.([0-9a-f]+) .*} bc-.* rx 131, edr -?[0-9]+ \((-?[0-9]+)\)$"
regex_dr_line = ".*TSCH: {asn-([0-9a-f])\.([0-9a-f]+) .*} bc-.* rx 131, dr -?[0-9]+ \((-?[0-9]+)\), edr -?[0-9]+ \((-?[0-9]+)\)$"

```
This will be added to the README.

**Comment by SoonSYJ** [comment]
Hi,
Thanks for your update. And I am confusing about the custom units (1/1024 of a microsecond). 

1. Is this means that a ~1GHz clock is used to measure the time? 
2. How can I understand the resolution of the synchronization error measurement which is figured in your paper (such as Fig. 9, Fig. 10 and Fig. 11)?
3. As I am using RTIMER_NOW() to print clock tick on Contiki, could do you give me some advice to get a high resolution synchronization measurement. 

Thanks a lot.

**Comment by atiselsts** [comment]
We describe the high-resolution time sync method in:
_A. Elsts, S. Duquennoy, X. Fafoutis, G. Oikonomou, R. Piechocki and I. Craddock, Microsecond-Accuracy Time Synchronization Using the IEEE 802.15.4 TSCH Protocol, IEEE LCN 2016._

The source code for it is also available: https://github.com/atiselsts/contiki/tree/ptsch
There are no external 1 GHz clocks involved. Only the MCU (`rtimer`) clock and the radio clock (`RAT`).

You might get by just using rtimers. The main problem is that the drift estimate will not be accurate. I have not tried using rtimers for this application, so cannot comment how much that would be a problem in practice - so it could be worth trying out! What we did try is to use rtimers for an application where sync happens more frequently and compare that with the high-precision approach. There the main problem is bigger sync errors, not the accuracy of the drift estimate (Fig. 11 in the usec accuracy paper).

**Comment by SoonSYJ** [comment]
Hi atiselsts,
As I am working on the original TSCH mac protocol from Contiki-OS, I have got some typical log message as showed below. The experiment is taken in an office environment without a chamber as your experiment. 

TSCH: {asn-0.10eba link-0-3-0-0 ch-20} drift -13
TSCH: {asn-0.10eba link-0-3-0-0 ch-20} uc-1-0 22 rx 76, dr -1, edr 1
TSCH: {asn-0.1104c link-0-3-0-0 ch-20} drift -12
TSCH: {asn-0.1104c link-0-3-0-0 ch-20} uc-1-0 22 rx 76, dr -4, edr 4

Taking look at the log message I have pasted, I have found that it is different from your log message for missing "dr -2 (-24562)". As your said that the value (-24562) is the "corrected synchronization error", I am not quite sure about the meaning of this value. I have tried to find the definition from the Contiki TSCH code, but failed. 

Could you please help me to printf this value from Contiki TSCH?
Thanks a lot.

**Comment by atiselsts** [comment]
Please, write new questions (if you have more) in a new issue.

On this last question: 

In your example, "dr -2 (-24562)", both `-2` and `-24562` are corrected synchronization errors. The first of those (`-2`) is registered by the standard Contiki TSCH time synchronization code and expressed in rtimer ticks (1/65536 of a second). The second (`-24562`) is registered by my high-precision sync code and expressed in high-precision units (1/1024 of microsecond). One rtimer tick is exactly 15625 high precision ticks long. In the example, `-24562 /  15625 = -1.571968`, which is approximately `-2`.

You can still use the rtimer tick values (`-2` in the example above, `dr -1` and `dr -4` in the log you pasted) to implement the same temperature compensation approach. But the results will not be as good. I don't know how good or bad exactly they will be as I haven't tried that.
